flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -78,19 +81,22 @@ public void startScheduling() {
 			deploymentOptions.put(schedulingVertex.getId(), option);
 		}
 
-		allocateSlotsAndDeployExecutionVertexIds(getAllVerticesFromTopology());
+		allocateSlotsAndDeployExecutionVertices(
+			getSchedulingExecutionVertices(getAllVerticesFromTopology()),
+			IS_IN_CREATED_STATE);
 	}
 
 	@Override
-	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+	public void restartTasks(final Set<ExecutionVertexID> verticesToRestart) {
+		final Set<SchedulingExecutionVertex> verticesToSchedule = getSchedulingExecutionVertices(verticesToRestart);
+
 		// increase counter of the dataset first
-		verticesToRestart
+		verticesToSchedule
 			.stream()
-			.map(schedulingTopology::getVertexOrThrow)
 			.flatMap(vertex -> vertex.getProducedResultPartitions().stream())
 			.forEach(inputConstraintChecker::resetSchedulingResultPartition);
 
-		allocateSlotsAndDeployExecutionVertexIds(verticesToRestart);
+		allocateSlotsAndDeployExecutionVertices(verticesToSchedule, IS_IN_TERMINAL_STATE);
```

Comment:
Can it happen at all that some of the vertices in `verticesToRestart` are not in a terminal state (`CANCELED`)?

Reply:
This should not happen because the `restartTasks` is only invoked when the `releaseFuture` of all the`verticesToRestart` are completed(valid states include `FINISHED`/`CANCELED`/`FAILED`).

Reply:
Would something speak against transitioning the _"vertices to restart"_ to `CREATED` already in the scheduler (before invoking `SchedulingStrategy#restartTasks()`)?

Reply:
A vertex is transitioned to `CREATED` only when it is reset. This only happens in `allocateSlotsAndDeploy` in `DefaultScheduler`.
The vertex versions will change in this case. 
And these vertices will be filtered out to restart.

Reply:
And I also think it's fine to not **restart** `CREATED` vertices. Some events later should be responsible to trigger the scheduling for them. Otherwise something might be wrong.

Reply:
> A vertex is transitioned to `CREATED` only when it is reset. This only happens in `allocateSlotsAndDeploy` in `DefaultScheduler`.
> The vertex versions will change in this case.
> And these vertices will be filtered out to restart.

I understand that we reset the vertices in `allocateSlotsAndDeploy`. However, if we reset the vertices before invoking `SchedulingStrategy#restartTasks()`, it looks to me that we don't need to distinguish in `allocateSlotsAndDeployExecutionVertices` between a restarted task and regular deployments.

Reply:
I see your concerns.
I'd also prefer to reset vertices before invoking `restartTasks()` on them. And the contract can be simplified that the schedulingStrategy should only schedule vertices which is in `CREATED` state.

Previously I thought we have made the decision to do lazy resetting. But now I think it's reasonable to change this behaviour.

Reply:
It just came to me that it can be problematic if we keep the lazy resetting.
e.g. The job A1--pipelined-->B1 fails. And only A1 will be re-scheduled on `restartTasks()` since the inputs of B1 are not ready. B1 should be scheduled later on the partition consumable event from restarted A1. But the terminal state of B1 will prevent B1 from being scheduled.

Reply:
And keeping a vertex in FAILED/CANCELED for a long time is also not good to users,  which can happen if it takes a long time for its inputs to become ready again.

How about closing FLINK-14248 and open a new ticket to change the resetting to happen before `restartTasks()`?

Reply:
Ok let's do it.

Reply:
> And only A1 will be re-scheduled on `restartTasks()` since the inputs of B1 are not ready. B1 should be scheduled later on the partition consumable event from restarted A1. But the terminal state of B1 will prevent B1 from being scheduled.

Nice catch.

Reply:
FLINK-14331 is opened to reset vertices right after they transition to terminated states.

Reply:
FLINK-14248 is closed.